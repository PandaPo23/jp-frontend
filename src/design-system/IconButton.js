import isNil from 'lodash/isNil';
import React from 'react';

import Flex from './Flex';
import Icon from './Icon';
import targetedProps from '../utils/targetedProps';
import Text from './Text';

const vert = (lp) => lp === 'top' || lp === 'bottom';
const before = (lp) => lp === 'left' || lp === 'top';

const IconButton = React.forwardRef(
  (
    {
      name,
      size = 24,
      bg = 'transparent',
      color = 'misc.muted',
      disabled = false,
      alignItems = 'center',
      border = 'none',
      label,
      children,
      labelPosition = 'right',
      labelMargin = 1,
      hoverable = true,
      justifyContent,
      labelUppercase = true,
      labelBold = true,
      as = 'button',
      ...props
    },
    ref
  ) => {
    const iconProps = targetedProps(props, 'icon');
    const labelProps = targetedProps(props, 'label');
    const justIcon = isNil(label) && !children;
    return (
      <Flex
        disabled={disabled}
        hoverable={hoverable}
        flexDirection={justIcon || vert(labelPosition) ? 'column' : 'row'}
        flexWrap="nowrap"
        nobr
        justifyContent={justifyContent || (justIcon && 'center')}
        alignItems={alignItems}
        border={border}
        color={color}
        bg={bg}
        as={as}
        ref={ref}
        {...props}
      >
        <Icon
          name={name}
          size={size}
          order={before(labelPosition) ? 2 : 1}
          {...iconProps}
        />
        {(children || !isNil(label)) && (
          <Text
            bg="transparent"
            disabled={disabled}
            order={before(labelPosition) ? 1 : 2}
            ml={labelPosition === 'right' && labelMargin}
            mr={labelPosition === 'left' && labelMargin}
            mt={labelPosition === 'bottom' && labelMargin}
            mb={labelPosition === 'top' && labelMargin}
            truncate
            nobr
            uppercase={labelUppercase}
            bold={labelBold}
            fontSize={3}
            {...labelProps}
          >
            {children || label}
          </Text>
        )}
      </Flex>
    );
  }
);

IconButton.displayName = 'IconButton';

export default IconButton;
