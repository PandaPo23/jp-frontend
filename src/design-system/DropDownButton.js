import React, { useEffect, useRef, useState } from 'react';
import { themeGet } from 'styled-system';
import { withTranslation } from 'react-i18next';
import { Manager, Reference, Popper } from 'react-popper';

import Box from './Box';
import Hr from './Hr';
import IconButton from './IconButton';
import Relative from './Relative';
import system from '../utils/System';
import Text from './Text';

export const MenuList = system('MenuList', {
  extend: Box,
  boxShadow: 2,
  overflow: 'hidden',
  bg: 'surface',
  zIndex: 10,
  borderRadius: 'sm',
});

export const ListItem = system(
  'ListItem',
  {
    extend: Text,
    px: 3,
    py: 3,
    width: 1,
    cursor: 'default',
  },
  (props) =>
    props.disabled
      ? {
          color: themeGet('colors.on.disabled')(props),
        }
      : {
          '&:hover': {
            backgroundColor: themeGet('colors.active')(props),
          },
        }
);

const useOutsideHandler = (ref, callback) => {
  const handleClickOutside = (event) => {
    if (ref.current && !ref.current.contains(event.target)) {
      callback();
    }
  };

  useEffect(() => {
    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  });
};

export const DropDownButton = ({
  label,
  items,
  value,
  onSelect,
  uppercase = false,
  iconSize = 28,
  t,
  customButtonRenderer,
  customMenuItemRenderer,
  menuMinWidth,
  placement = 'auto-end',
  ...props
}) => {
  const [open, setOpen] = useState(false);
  const wrapperRef = useRef(null);
  const activeItem = items.find((item) => item.value === value) || {};
  const [focusedIndex, setFocusedIndex] = useState(-1);

  useOutsideHandler(wrapperRef, () => setOpen(false));

  const navigateDropDown = (event, items) => {
    let key = event.keyCode;
    let newIndex = 0;
    let item = items.find((item, index) => index === focusedIndex);
    switch (key) {
      // press enter
      case 13:
        if (item && !item.disabled) {
          handleToggleAndSelect(item);
        }
        return;
      // press ESC
      case 27:
        return setOpen(false);
      // press ARROW UP
      case 40:
        if (focusedIndex === items.length) {
          setFocusedIndex(0);
          return;
        }
        newIndex = focusedIndex + 1;
        break;
      // press ARROW UP
      case 38:
        if (focusedIndex === 0) {
          setFocusedIndex(items.length);
          return;
        }
        newIndex = focusedIndex - 1;
        break;
      default:
        break;
    }
    setFocusedIndex(newIndex);
  };

  const handleToggleAndSelect = (item, type = 'navigationClick') => {
    onSelect && onSelect(item);
    if (type !== 'navigationClick') {
      setOpen(false);
    }
    setFocusedIndex(-1);
  };

  const handleClick = () => setOpen(open ? false : true);
  const textStyle = uppercase ? 'uppercase' : 'normal';

  const customRendererProps = {
    onClick: handleClick,
    iconSize,
    uppercase,
    open,
    item: activeItem,
  };

  return (
    <Relative
      {...props}
      onKeyDown={(e) => navigateDropDown(e, items)}
      ref={wrapperRef}
    >
      <Manager>
        <Reference>
          {({ ref }) =>
            customButtonRenderer ? (
              customButtonRenderer({ ...customRendererProps, ref })
            ) : (
              <IconButton
                labelPosition="left"
                ref={ref}
                onClick={handleClick}
                size={iconSize}
                name={open ? 'arrow-drop-up' : 'arrow-drop-down'}
                px={1}
                textStyle={textStyle}
                label={activeItem.value ? activeItem.label : label}
              />
            )
          }
        </Reference>
        <Popper placement={placement} eventsEnabled>
          {({ ref, style, placement, arrowProps }) =>
            open ? (
              <MenuList
                ref={ref}
                style={style}
                data-placement={placement}
                minWidth={menuMinWidth || undefined}
              >
                {items.map((item, index) => {
                  const selected = activeItem.value === item.value;
                  const focus = index === focusedIndex;
                  const { label, disabled } = item;
                  return item.type === 'divider' ? (
                    <Hr key={index} my={2} />
                  ) : (
                    <ListItem
                      key={index}
                      active={
                        customMenuItemRenderer
                          ? focus
                          : focus || (focusedIndex === -1 && selected)
                      }
                      disabled={disabled}
                      onClick={() =>
                        !disabled && handleToggleAndSelect(item, 'itemClick')
                      }
                    >
                      {customMenuItemRenderer
                        ? customMenuItemRenderer({
                            label,
                            selected,
                            disabled: item.disabled,
                          })
                        : label}
                    </ListItem>
                  );
                })}
              </MenuList>
            ) : null
          }
        </Popper>
      </Manager>
    </Relative>
  );
};

export default withTranslation('common')(DropDownButton);
