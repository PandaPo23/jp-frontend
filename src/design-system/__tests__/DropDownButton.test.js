import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { DropDownButton, MenuList, ListItem } from '../DropDownButton';
import IconButton from '../IconButton';
import Relative from '../Relative';

const items = [
  {
    value: 'byName',
    label: 'by Name',
  },
  {
    value: 'byDate',
    label: 'by Date',
  },
  {
    value: 'byAge',
    label: 'by Age',
  },
];

describe('DropDownButton component', () => {
  it('renders correctly', () => {
    const localProps = {
      t: (t) => t,
      label: 'Sort by',
      color: 'on.surface',
      value: null,
      onSelect: () => {},
      items,
    };
    const wrapper = mount(<DropDownButton {...localProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find(IconButton).prop('label')).toEqual(localProps.label);
  });

  it('opens or hides the dropdown menu with user interaction', () => {
    const localProps = {
      t: (t) => t,
      label: 'Sort by',
      color: 'on.surface',
      value: 'byName',
      onSelect: jest.fn(),
      items,
    };
    const wrapper = mount(<DropDownButton {...localProps} />);

    expect(wrapper.find(MenuList)).toHaveLength(0);

    wrapper.find(IconButton).simulate('click');
    expect(wrapper.find(MenuList)).toHaveLength(1);

    wrapper.find(Relative).simulate('keydown', { keyCode: 13 });
    expect(wrapper.find(MenuList)).toHaveLength(1);

    wrapper.find(Relative).simulate('keydown', { keyCode: 13 });
    wrapper.find(Relative).simulate('keydown', { keyCode: 27 });
    expect(wrapper.find(MenuList)).toHaveLength(0);
  });

  it('triggers event handler prop functions when dropdown item is selected', () => {
    const localProps = {
      t: (t) => t,
      label: 'Sort by',
      color: 'on.surface',
      value: 'byName',
      onSelect: jest.fn(),
      items,
    };
    const wrapper = mount(<DropDownButton {...localProps} />);
    wrapper.find(IconButton).simulate('click');
    wrapper
      .find(MenuList)
      .find(ListItem)
      .at(0)
      .simulate('click');
    expect(localProps.onSelect).toHaveBeenCalledWith(items[0]);
  });

  it('on keydown select item', () => {
    const localProps = {
      t: (t) => t,
      label: 'Sort by',
      color: 'on.surface',
      value: 'byDate',
      onSelect: jest.fn(),
      items,
    };
    const wrapper = mount(<DropDownButton {...localProps} />);
    wrapper
      .find(Relative)
      .simulate('keydown', { keyCode: 13 })
      .simulate('keydown', { keyCode: 40 })
      .simulate('keydown', { keyCode: 40 })
      .simulate('keydown', { keyCode: 13 });
    expect(localProps.onSelect).toHaveBeenCalledWith(items[1]);
    expect(wrapper.find(MenuList)).toHaveLength(0);
  });
});
