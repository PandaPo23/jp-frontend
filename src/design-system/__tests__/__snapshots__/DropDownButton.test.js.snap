// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDownButton component renders correctly 1`] = `
.emotion-6 {
  color: on.surface;
  position: relative;
  position: relative;
}

.emotion-5 {
  color: misc.muted;
  background-color: transparent;
  padding-left: 4px;
  padding-right: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  border: none;
  cursor: pointer;
}

.emotion-0 {
  min-width: 28px;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-order: 2;
  -ms-flex-order: 2;
  order: 2;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 28px;
}

.emotion-4 {
  background-color: transparent;
  font-size: 20px;
  margin-right: 4px;
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
  font-weight: bold;
  text-transform: uppercase;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

<DropDownButton
  color="on.surface"
  items={
    Array [
      Object {
        "label": "by Name",
        "value": "byName",
      },
      Object {
        "label": "by Date",
        "value": "byDate",
      },
      Object {
        "label": "by Age",
        "value": "byAge",
      },
    ]
  }
  label="Sort by"
  onSelect={[Function]}
  t={[Function]}
  value={null}
>
  <Relative
    color="on.surface"
    onKeyDown={[Function]}
    position="relative"
  >
    <div
      className="emotion-6"
      onKeyDown={[Function]}
    >
      <Manager>
        <Reference>
          <InnerReference
            setReferenceNode={[Function]}
          >
            <IconButton
              label="Sort by"
              labelPosition="left"
              name="arrow-drop-down"
              onClick={[Function]}
              px={1}
              size={28}
              textStyle="normal"
            >
              <Flex
                alignItems="center"
                as="button"
                bg="transparent"
                border="none"
                color="misc.muted"
                disabled={false}
                display="flex"
                flexDirection="row"
                flexWrap="nowrap"
                hoverable={true}
                justifyContent={false}
                nobr={true}
                onClick={[Function]}
                px={1}
                textStyle="normal"
              >
                <button
                  className="emotion-5"
                  disabled={false}
                  onClick={[Function]}
                >
                  <Icon
                    name="arrow-drop-down"
                    order={2}
                    size={28}
                  >
                    <Base
                      className="emotion-0"
                      name="arrow-drop-down"
                      order={2}
                      size={28}
                    >
                      <Flex
                        centered={true}
                        className="emotion-0"
                        display="flex"
                        order={2}
                      >
                        <div
                          className="emotion-1"
                        >
                          <svg
                            className="emotion-0"
                            name="arrow-drop-down"
                            order={2}
                            size={28}
                          />
                        </div>
                      </Flex>
                    </Base>
                  </Icon>
                  <Text
                    bg="transparent"
                    bold={true}
                    disabled={false}
                    fontSize={3}
                    mb={false}
                    ml={false}
                    mr={1}
                    mt={false}
                    nobr={true}
                    order={1}
                    truncate={true}
                    uppercase={true}
                  >
                    <div
                      className="emotion-4"
                      disabled={false}
                    >
                      Sort by
                    </div>
                  </Text>
                </button>
              </Flex>
            </IconButton>
          </InnerReference>
        </Reference>
        <Popper
          eventsEnabled={true}
          placement="auto-end"
        >
          <InnerPopper
            eventsEnabled={true}
            placement="auto-end"
            positionFixed={false}
            referenceElement={
              .emotion-3 {
  color: misc.muted;
  background-color: transparent;
  padding-left: 4px;
  padding-right: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  border: none;
  cursor: pointer;
}

.emotion-0 {
  min-width: 28px;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-order: 2;
  -ms-flex-order: 2;
  order: 2;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 28px;
}

.emotion-2 {
  background-color: transparent;
  font-size: 20px;
  margin-right: 4px;
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
  font-weight: bold;
  text-transform: uppercase;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

<button
                class="emotion-3"
              >
                <div
                  class="emotion-1"
                >
                  <svg
                    class="emotion-0"
                    name="arrow-drop-down"
                    order="2"
                    size="28"
                  />
                </div>
                <div
                  class="emotion-2"
                >
                  Sort by
                </div>
              </button>
            }
          />
        </Popper>
      </Manager>
    </div>
  </Relative>
</DropDownButton>
`;
